src/btree/bt_sync_obsolete.c

```diff
@@ -291,3 +290,325 @@ __wt_sync_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *parent)
 
     return (0);
 }
+
+/*
+ * __checkpoint_cleanup_run_chk --
+ *     Check to decide if the checkpoint cleanup should continue running.
+ */
+static bool
+__checkpoint_cleanup_run_chk(WT_SESSION_IMPL *session)
+{
+    return (FLD_ISSET(S2C(session)->server_flags, WT_CONN_SERVER_CHECKPOINT_CLENAUP));
+}
+
+/*
+ * __checkpoint_cleanup_page_skip --
+ *     Return if checkpoint cleanup requires we read this page.
+ */
+static int
+__checkpoint_cleanup_page_skip(
+  WT_SESSION_IMPL *session, WT_REF *ref, void *context, bool visible_all, bool *skipp)
+{
+    WT_ADDR_COPY addr;
+
+    WT_UNUSED(context);
+    WT_UNUSED(visible_all);
+
+    *skipp = false; /* Default to reading */
+
+    /*
+     * Skip deleted pages as they are no longer required for the checkpoint. The checkpoint never
+     * needs to review the content of those pages - if they should be included in the checkpoint the
+     * existing page on disk contains the right information and will be linked into the checkpoint
+     * as the internal tree structure is built.
+     */
+    if (ref->state == WT_REF_DELETED) {
+        *skipp = true;
+        return (0);
+    }
+
+    /* If the page is in-memory, we want to look at it. */
+    if (ref->state != WT_REF_DISK)
+        return (0);
+
+    /*
+     * Reading any page that is not in the cache will increase the cache size. Perform a set of
+     * checks to verify the cache can handle it.
+     */
+    if (__wt_cache_aggressive(session) || __wt_cache_full(session) || __wt_cache_stuck(session) ||
+      __wt_eviction_needed(session, false, false, NULL)) {
+        *skipp = true;
+        return (0);
+    }
+
+    /* Don't read pages into cache during startup or shutdown phase. */
+    if (F_ISSET(S2C(session), WT_CONN_RECOVERING | WT_CONN_CLOSING_CHECKPOINT)) {
+        *skipp = true;
+        return (0);
+    }
+
+    /*
+     * Ignore the pages with no on-disk address. It is possible that a page with deleted state may
+     * not have an on-disk address.
+     */
+    if (!__wt_ref_addr_copy(session, ref, &addr))
+        return (0);
+
+    /*
+     * The checkpoint cleanup fast deletes the obsolete leaf page by marking it as deleted
```

Comment:
Should we still call it checkpoint cleanup then?

Reply:
Hmm, we can change the name also. Any suggestions?

Reply:
What about obsolete_cleanup or garbage_cleanup?

Reply:
When this feature was first implemented, we decided not to go with `garbage cleanup`. `obsolete_cleanup` is fine. I will leave the name change to another PR.